ios/AirGoogleMaps/AIRGoogleMap.m

```diff
@@ -0,0 +1,222 @@
+
+- (id)eventFromCoordinate:(CLLocationCoordinate2D)coordinate;
+
+@end
+
+@implementation AIRGoogleMap
+{
+  NSMutableArray<UIView *> *_reactSubviews;
+  BOOL _initialRegionSet;
+}
+
+- (instancetype)init
+{
+  if ((self = [super init])) {
+    _reactSubviews = [NSMutableArray new];
+    _markers = [NSMutableArray array];
+    _initialRegionSet = false;
+  }
+  return self;
+}
+- (id)eventFromCoordinate:(CLLocationCoordinate2D)coordinate {
+
+  CGPoint touchPoint = [self.projection pointForCoordinate:coordinate];
+
+  return @{
+           @"coordinate": @{
+               @"latitude": @(coordinate.latitude),
+               @"longitude": @(coordinate.longitude),
+               },
+           @"position": @{
+               @"x": @(touchPoint.x),
+               @"y": @(touchPoint.y),
+               },
+           };
+}
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wobjc-missing-super-calls"
+- (void)insertReactSubview:(id<RCTComponent>)subview atIndex:(NSInteger)atIndex {
+  // Our desired API is to pass up markers/overlays as children to the mapview component.
+  // This is where we intercept them and do the appropriate underlying mapview action.
+  if ([subview isKindOfClass:[AIRGoogleMapMarker class]]) {
+    AIRGoogleMapMarker *marker = (AIRGoogleMapMarker*)subview;
+    marker.realMarker.map = self;
+    [self.markers addObject:marker];
+  }
+  [_reactSubviews insertObject:(UIView *)subview atIndex:(NSUInteger) atIndex];
+}
+#pragma clang diagnostic pop
+
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wobjc-missing-super-calls"
+- (void)removeReactSubview:(id<RCTComponent>)subview {
+  // similarly, when the children are being removed we have to do the appropriate
+  // underlying mapview action here.
+  if ([subview isKindOfClass:[AIRGoogleMapMarker class]]) {
+    AIRGoogleMapMarker *marker = (AIRGoogleMapMarker*)subview;
+    marker.realMarker.map = nil;
+    [self.markers removeObject:marker];
+  }
+  [_reactSubviews removeObject:(UIView *)subview];
+}
+#pragma clang diagnostic pop
+
+- (void)setInitialRegion:(MKCoordinateRegion)initialRegion {
+  if (_initialRegionSet) return;
+
+  _initialRegion = initialRegion;
+  _initialRegionSet = true;
+
+  float latitudeDelta = initialRegion.span.latitudeDelta * 0.5;
+  float longitudeDelta = initialRegion.span.longitudeDelta * 0.5;
+
+  CLLocationCoordinate2D a = CLLocationCoordinate2DMake(initialRegion.center.latitude + latitudeDelta,
+                                                         initialRegion.center.longitude + longitudeDelta);
+  CLLocationCoordinate2D b = CLLocationCoordinate2DMake(initialRegion.center.latitude - latitudeDelta,
+                                                         initialRegion.center.longitude - longitudeDelta);
+  GMSCoordinateBounds *bounds = [[GMSCoordinateBounds alloc] initWithCoordinate:a coordinate:b];
+  GMSCameraPosition *camera = [self cameraForBounds:bounds insets:UIEdgeInsetsZero];
```

Comment:
it really seems like we should have a pulled out and reusable function to convert an `MKCoordinateRegion` to a `GMSCoordinateBounds`